### PR TITLE
feat: companion extension

### PR DESCRIPTION
**Pull request by rebelchris**
## Changes

- Added the companion application
- Added basic companion render (chrome) test cases as the individual actions are all existing implementations.

## Manual Testing

- On those affected packages:
- [x] Have you done sanity checks in the webapp?
- [x] Have you done sanity checks in the extension?

- Did you test the modified components media queries?
- [x] MobileL (420px) (Not super applicable, but it works)
- [x] Tablet (656px)
- [x] Laptop (1020px)

DD-458 #done

### CODE REVIEW

File: packages/extension/package.json [hunk 1]
```diff
@@ -36,6 +36,7 @@
     "react-intersection-observer": "^8.32.2",
     "react-modal": "^3.14.3",
     "react-query": "^3.31.0",
+    "react-shadow-root": "^6.1.1",
```
Comment (rebelchris) [inline packages/extension/package.json:39]: Not sure if we actually need this anymore

File: packages/extension/package.json [hunk 2]
```diff
@@ -110,9 +111,11 @@
     "postcss-loader": "^4.0.4",
     "postcss-mixins": "^8.1.0",
     "postcss-nesting": "^7.0.1",
+    "postcss-rem-to-responsive-pixel": "^5.1.1",
     "postcss-simple-vars": "^6.0.3",
     "prettier": "^2.4.1",
     "resolve-url-loader": "^3.1.4",
+    "@open-wc/testing": "^2.1.2",
```
Comment (rebelchris) [inline packages/extension/package.json:118]: Double check this one as well

File: packages/extension/src/background/index.ts [hunk 3]
```diff
@@ -15,6 +27,95 @@ const cacheAmplitudeDeviceId = async ({
   }
 };
 
+const excludedCompanionOrigins = [
+  'https://twitter.com',
+  'https://www.google.com',
+  'chrome-extension://',
+];
```
Comment (rebelchris) [inline packages/extension/src/background/index.ts:34]: This is kind of a first draft of hard-coded domains we would never want to fire on.
We can extend this or make this work differently in the future.

The idea was to have sites that devs use a lot, but never are featured blacklisted from Querying the API

File: packages/extension/src/companion/App.tsx [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    id: 'companion-options-context',
+  });
+  const onContextOptions = (event: React.MouseEvent): void => {
+    showCompanionOptionsMenu(event, {
+      position: { x: 48, y: 265 },
+    });
+  };
+
+  return (
+    <div
```
Comment (rebelchris) [inline packages/extension/src/companion/App.tsx:100]: Perhaps we can refactor this into multiple smaller components

File: packages/shared/src/styles/base.css [hunk 5]
```diff
@@ -1,8 +1,15 @@
-html {
+html, #daily-companion-wrapper {
```
Comment (rebelchris) [inline packages/shared/src/styles/base.css:1]: Check if we still use this one please

File: packages/shared/src/styles/base.css [hunk 6]
```diff
@@ -21,6 +28,10 @@ body {
   &.ReactModal__Body--open .ReactModal__Content > * {
     overscroll-behavior-y: none;
   }
+   &.companion {
+  min-width: auto;
+  overflow: hidden;
+    }
```
Comment (rebelchris) [inline packages/shared/src/styles/base.css:34]: Thinks this can be removed as well

**Review by rebelchris — COMMENTED** [review]
Some small notes

File: packages/extension/src/background/index.ts (hunk 3 above)
Comment (idoshamun) [reply packages/extension/src/background/index.ts:34]: We can add stackoverflow, gmail, google meet, google calendar. Don't know if it makes sense

File: packages/extension/public/css/daily-companion-app.css [hunk 7]
```diff
@@ -0,0 +1,8 @@
+daily-companion-app {
+  line-height: 1.5; /* 1 */
```
Comment (sshanzel) [inline packages/extension/public/css/daily-companion-app.css:2]: Just a question, what does the number comments signify?

File: packages/extension/src/background/index.ts [hunk 8]
```diff
@@ -15,6 +27,95 @@ const cacheAmplitudeDeviceId = async ({
...
+
+const isExcluded = (origin: string) =>
+  excludedCompanionOrigins.some((e) => origin.includes(e));
+
+const sendBootData = async (req, sender) => {
+  if (isExcluded(sender?.origin)) {
+    return;
+  }
+  const cacheData = JSON.parse(localStorage.getItem('boot:local'));
```
Comment (sshanzel) [inline packages/extension/src/background/index.ts:43]: I think we have a generic function that fetches the local boot data.

File: packages/extension/src/companion/CompanionContextMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+  onReport,
+  onBlockSource,
+  onDisableCompanion,
+}: {
+  postData: BootData;
+  onMessage?: (message: string, timeout?: number) => Promise<unknown>;
+  onReport;
+  onBlockSource;
+  onDisableCompanion;
+}): ReactElement {
```
Comment (sshanzel) [inline packages/extension/src/companion/CompanionContextMenu.tsx:24]: I think we can create an interface for this so we can also explicitly define the parameters.

File: packages/shared/src/lib/boot.ts [hunk 10]
```diff
@@ -13,17 +13,24 @@ export type Boot = {
...
   flags: IFlags;
   settings: RemoteSettings;
+  data?;
 };
 
 export type BootCacheData = Pick<
   Boot,
-  'user' | 'alerts' | 'settings' | 'flags'
->;
+  'user' | 'alerts' | 'settings' | 'flags' | 'data'
```
Comment (sshanzel) [inline packages/shared/src/lib/boot.ts:21]: What is the content of `data` property? I think we should change it to something much more explicit.

File: packages/shared/src/components/modals/ReportPostModal.tsx [hunk 11]
```diff
@@ -14,7 +14,7 @@ import {
 
 export interface Props extends ModalProps {
   postIndex: number;
-  post: Post;
+  post?: Partial<Post>;
```
Comment (sshanzel) [inline packages/shared/src/components/modals/ReportPostModal.tsx:17]: Why do we have to make this as a `Partial`type? I think the post should be required to report something.

File: packages/extension/src/companion/common.tsx [hunk 12]
```diff
@@ -0,0 +1,20 @@
+import { Post } from '@dailydotdev/shared/src/graphql/posts';
+
+export type BootData = Pick<
```
Comment (sshanzel) [inline packages/extension/src/companion/common.tsx:3]:
```suggestion
export type PostBootData = Pick<
```

File: packages/extension/src/companion/useCompanionActions.ts [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+  removeUpvote: (variables: T) => Promise<void>;
+  disableCompanion: (variables: T) => Promise<void>;
+};
+
+export default function useCompanionActions<
+  T extends { id?: string; reason?: string; comment?: string } = {
+    id?: string;
+    reason?: string;
+    comment?: string;
+  },
```
Comment (sshanzel) [inline packages/extension/src/companion/useCompanionActions.ts:26]: I think we should create interfaces for this, to make it easier to understand.

**Review by sshanzel — CHANGES_REQUESTED** [review]
Wow, I can't imagine the amount of work you did put into this. Must have been crazy moments 💪 💪 💪 

Just had a few comments and questions but nothing major. Super amazing work you got here, Chris! 🎸 🎸 🎸

File: packages/shared/src/components/modals/ReportPostModal.tsx (hunk 11 above)
Comment (rebelchris) [reply packages/shared/src/components/modals/ReportPostModal.tsx:17]: It should indeed, I must actually remake it to be type of CompanionPost/Post
As the CompanionPost has less data compared to a full post.

File: packages/extension/public/css/daily-companion-app.css (hunk 7 above)
Comment (rebelchris) [reply packages/extension/public/css/daily-companion-app.css:2]: To be completely honest, no idea.
It was something we had setup in the extension, and was thinking it might have been a fallback for certain old browsers?

File: packages/extension/src/background/index.ts (hunk 3 above)
Comment (idoshamun) [reply packages/extension/src/background/index.ts:34]: Following up on our call yesterday, I see that `excludedCompanionOrigins` is your own implementation and not something we define on the manifest level. So it's great to have this list and when needed we can just remove sites from this list if we suddenly want to support them

File: packages/extension/src/background/index.ts [hunk 14]
```diff
@@ -1,5 +1,17 @@
...
+  REMOVE_BOOKMARK_MUTATION,
+  REPORT_POST_MUTATION,
+  UPVOTE_MUTATION,
+} from '@dailydotdev/shared/src/graphql/posts';
+import request from 'graphql-request';
+import { ADD_FILTERS_TO_FEED_MUTATION } from '@dailydotdev/shared/src/graphql/feedSettings';
+import { UPDATE_USER_SETTINGS_MUTATION } from '@dailydotdev/shared/src/graphql/settings';
+import { BOOT_LOCAL_KEY } from '@dailydotdev/shared/src/contexts/BootProvider';
 
 const cacheAmplitudeDeviceId = async ({
```
Comment (idoshamun) [inline packages/extension/src/background/index.ts:16]: This is no longer needed I think.
It should have been removed during the third party tracking task but we probably forgot

File: packages/extension/src/background/index.ts (hunk 8 above)
Comment (idoshamun) [reply packages/extension/src/background/index.ts:43]: Not sure we need the cached data, to be honest. Because the companion boot is a requirement before we can render anything. So when we need to render something we already have the most up-to-date boot data available.

File: packages/extension/src/background/index.ts [hunk 15]
```diff
@@ -15,6 +27,95 @@ const cacheAmplitudeDeviceId = async ({
...
+  });
+};
+
+function handleMessages(message, sender) {
+  if (message.type === 'CONTENT_LOADED') {
+    sendBootData(message, sender);
+    return null;
+  }
+
+  if (message.type === 'UPVOTE_POST') {
```
Comment (idoshamun) [inline packages/extension/src/background/index.ts:61]: A reminder that we want to change the API with the background worker to be something like a proxy to the `request` function rather than implementing a new type for every graphql request.

File: packages/extension/src/companion/index.tsx [hunk 16]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import { browser } from 'webextension-polyfill-ts';
+import { themeModes } from '@dailydotdev/shared/src/contexts/SettingsContext';
+import { CompanionBootData } from './common';
+import App from './App';
+
+const init = async (data: CompanionBootData, settings) => {
```
Comment (idoshamun) [inline packages/extension/src/companion/index.tsx:8]: I suggest a more explicit name like `render` or `renderApp`

File: packages/extension/src/background/index.ts [hunk 17]
```diff
@@ -15,6 +27,95 @@ const cacheAmplitudeDeviceId = async ({
...
+  }
+  await browser.tabs.sendMessage(sender?.tab?.id, {
+    data,
+    settings: settingsOutput,
+  });
+};
+
+function handleMessages(message, sender) {
+  if (message.type === 'CONTENT_LOADED') {
+    sendBootData(message, sender);
```
Comment (idoshamun) [inline packages/extension/src/background/index.ts:57]: If the user has opted out we should not send a request to our server at all, I think. What do you think?

File: packages/extension/src/companion/App.tsx [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+  if (isOptOutCompanion) {
+    return <></>;
+  }
+
+  return (
+    <div>
+      <style>
+        @import &quot;chrome-extension://{browser.runtime.id}
+        /css/companion.css&quot;;
+      </style>
```
Comment (idoshamun) [inline packages/extension/src/companion/App.tsx:242]: Is there no other way to achieve it? How do we do it in the extension?

File: packages/extension/src/companion/App.tsx (hunk 4 above)
Comment (idoshamun) [reply packages/extension/src/companion/App.tsx:100]: I agree this is too big and hard to read

File: packages/extension/src/companion/CompanionContextMenu.tsx [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+          url: shareLink,
+        });
+      } catch (err) {
+        // Do nothing
+      }
+    } else {
+      await navigator.clipboard.writeText(shareLink);
+      onMessage('✅ Copied link to clipboard');
+    }
+  };
```
Comment (idoshamun) [inline packages/extension/src/companion/CompanionContextMenu.tsx:43]: Don't we have a function like this already?

File: packages/extension/src/companion/CompanionContextMenu.tsx [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+          <EyeIcon className="mr-2 text-xl" /> Disable widget
+        </Item>
+      </Menu>
+      {reportModal && (
+        <RepostPostModal
+          post={postData}
+          parentSelector={() =>
+            document
+              .querySelector('daily-companion-app')
+              .shadowRoot.querySelector('#daily-companion-wrapper')
```
Comment (idoshamun) [inline packages/extension/src/companion/CompanionContextMenu.tsx:97]: I suggest turning it into a function

File: packages/shared/src/components/modals/LoginModal.tsx [hunk 21]
```diff
@@ -34,7 +33,6 @@ export default function LoginModal({
       className={classNames(styles.loginModal, className)}
     >
       <ModalCloseButton onClick={onRequestClose} />
-      <DailyDevLogo />
```
Comment (idoshamun) [inline packages/shared/src/components/modals/LoginModal.tsx:37]: Why did you remove the logo?

File: packages/shared/src/contexts/BootProvider.tsx [hunk 22]
```diff
@@ -70,12 +70,13 @@ const updateLocalBootData = (
   current: Partial<BootCacheData>,
   boot: Partial<BootCacheData>,
 ) => {
-  const localData = { ...current, ...boot };
+  const localData = { ...current, ...boot, last_modifier: 'extension' };
```
Comment (idoshamun) [inline packages/shared/src/contexts/BootProvider.tsx:73]: Why not camelCase?

File: packages/shared/src/contexts/BootProvider.tsx [hunk 23]
```diff
@@ -112,20 +113,20 @@ export const BootDataProvider = ({
...
     } else {
-      if (lastAppliedChange) {
+      if (cachedData?.last_modifier !== 'companion' && lastAppliedChange) {
         updatedData = { ...updatedData, ...lastAppliedChange };
       }
       setLastAppliedChange(null);
     }
-
-    const updated = updateLocalBootData(cachedBootData, updatedData);
+    const updated = updateLocalBootData(cachedData, updatedData);
```
Comment (idoshamun) [inline packages/shared/src/contexts/BootProvider.tsx:129]: Not sure I understand the reason for this change

File: packages/shared/src/styles/base.css [hunk 24]
```diff
@@ -1,8 +1,15 @@
...
   color: var(--theme-label-primary);
   height: -webkit-fill-available;
 }
+daily-companion-app {
+  line-height: 1.5; /* 1 */
+  -webkit-text-size-adjust: 100%; /* 2 */
+  -moz-tab-size: 4; /* 3 */
+  tab-size: 4; /* 3 */
+  font-family: theme('fontFamily.sans', ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji"); /* 4 */
+}
```
Comment (idoshamun) [inline packages/shared/src/styles/base.css:12]: If you copied it from somewhere else maybe we can utilize the same selector?

File: packages/extension/public/css/daily-companion-app.css (hunk 7 above)
Comment (sshanzel) [reply packages/extension/public/css/daily-companion-app.css:2]: Oh, cool. Thought this was something new. Thank you!

**Review by idoshamun — CHANGES_REQUESTED** [review]
I did a high level review looked for some big chunks and put my comments.
You overcame so many challenges in this single PR that it is definitely amazing 🤩 
Can't wait to see the impact of it

File: packages/extension/src/background/index.ts (hunk 8 above)
Comment (rebelchris) [reply packages/extension/src/background/index.ts:43]: This part is only for anonymous user syncing if they modify the theme in the extension and move over to a companion enabled website.

File: packages/extension/src/background/index.ts (hunk 17 above)
Comment (rebelchris) [reply packages/extension/src/background/index.ts:57]: The main issue is that this info only comes from the boot query.
We could do it based on the local cache, but it might not always be the latest value.

File: packages/extension/src/companion/App.tsx (hunk 18 above)
Comment (rebelchris) [reply packages/extension/src/companion/App.tsx:242]: The main issue is actually shadow dom not being able to access regular imports in React.
I can have another look at it though.

File: packages/shared/src/contexts/BootProvider.tsx (hunk 23 above)
Comment (rebelchris) [reply packages/shared/src/contexts/BootProvider.tsx:129]: Thanks for raising this.
We introduced the lastAppliedChange check which could favorite local data over remote data.
however with the introduction of the companion there is a "third" race condition.

When the companion runs a update we always want to favorite that, (it's the only way this last_modifier get's set to companion)

Else we got into edge cases where you turn of the companion (from the companion)
And the extension would re-enable it again from cache or remote data.

Let me know if that makes sense, happy to showcase it if that helps

File: packages/extension/src/background/index.ts (hunk 8 above)
Comment (idoshamun) [reply packages/extension/src/background/index.ts:43]: I see ok then...

File: packages/extension/src/background/index.ts (hunk 17 above)
Comment (idoshamun) [reply packages/extension/src/background/index.ts:57]: I'm ok with this tradeoff. If the user opted out we should reduce the overhead of this extension to minimum

File: packages/extension/src/companion/App.tsx (hunk 18 above)
Comment (idoshamun) [reply packages/extension/src/companion/App.tsx:242]: Actually, I'm thinking of a webpack trick perhaps. The best way would be to inline the CSS to there during the build. Not sure if it's easy to achieve though.

File: packages/shared/src/contexts/BootProvider.tsx (hunk 23 above)
Comment (idoshamun) [reply packages/shared/src/contexts/BootProvider.tsx:129]: We need to document it somehow. It will be very hard to maintain this piece without understanding the side effects

File: packages/extension/src/background/index.ts (hunk 14 above)
Comment (rebelchris) [reply packages/extension/src/background/index.ts:19]: Refactoring in separate PR

File: packages/shared/src/contexts/BootProvider.tsx (hunk 23 above)
Comment (rebelchris) [reply packages/shared/src/contexts/BootProvider.tsx]: Let me create a DR for this and describe the flow we choose

File: packages/extension/src/companion/App.tsx (hunk 18 above)
Comment (rebelchris) [reply packages/extension/src/companion/App.tsx:31]: I tried some of the known webpack tips:
- import directly -> doesn't build in the shadow dom
- import as variable and inject output -> outputs empty

What would be the benefit of doing that vs the import like this?
As that companion.css gets dynamically build for every build. (It's a packed version of the global.css)

Comment (idoshamun) [reply packages/extension/src/companion/App.tsx:30]: To streamline the way we use global css files but don't put any more time into it. It's not that important

File: packages/extension/src/companion/companionRequest.tsx [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+      type: 'GRAPHQL_REQUEST',
+      url: args?.[0],
+      document: args?.[1],
+      variables: args?.[2],
+    });
+    return null;
+  },
+};
+
+export const companionRequest = new Proxy(request, proxyRequest);
```
Comment (idoshamun) [inline packages/extension/src/companion/companionRequest.tsx:16]: What's the advantage of using `Proxy`?
Because it seems like you must import the `graphql-request` module here although you don't really use it

**Review by idoshamun — CHANGES_REQUESTED** [review]
Looks great only one question

File: packages/extension/src/companion/companionRequest.tsx (hunk 25 above)
Comment (rebelchris) [reply packages/extension/src/companion/companionRequest.tsx:16]: I think the main advantage is that is copies the signature so we don't have to explicitly state those, and we can simply "modify" what it actually does.

Would you rather have a function that just has the same props?

Comment (idoshamun) [reply packages/extension/src/companion/companionRequest.tsx:16]: I would favor a solution that doesn't involve importing unnecessary modules in this case

**Review by sshanzel — APPROVED** [review]
Sharp work! Nothing to add from my end. Can't believe we are nearing the merge of this feature. Great work, no doubt! 🚢

File: packages/extension/src/companion/App.tsx [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+              visit={visit}
+              tokenRefreshed
+              getRedirectUri={() => '/'}
+              updateUser={() => null}
+            >
+              <SettingsContextProvider settings={settings}>
+                <AlertContextProvider alerts={alerts}>
+                  <AnalyticsContextProvider
+                    app="companion"
+                    version="1"
```
Comment (idoshamun) [inline packages/extension/src/companion/App.tsx:59]: Can you provide the real extension version here, please?

File: packages/extension/src/companion/App.tsx [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+
+  return (
+    <div>
+      <style>
+        @import &quot;chrome-extension://{browser.runtime.id}
+        /css/companion.css&quot;;
+      </style>
+      <RouterContext.Provider value={router}>
+        <QueryClientProvider client={queryClient}>
+          <FeaturesContext.Provider value={{ flags }}>
```
Comment (idoshamun) [inline packages/extension/src/companion/App.tsx:47]: Usually, it's better to have `{{ flags }}` memoized. Otherwise it creates a new object on every render

File: packages/extension/src/companion/index.tsx [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+
+const renderApp = ({ ...props }: CompanionData) => {
+  const { settings } = props;
+  getCompanionWrapper().classList.add(themeModes[settings.theme]);
+
+  // Set target of the React app to shadow dom
+  ReactDOM.render(<App {...props} />, getCompanionWrapper());
+};
+
+browser.runtime.onMessage.addListener(async (request) => {
```
Comment (idoshamun) [inline packages/extension/src/companion/index.tsx:16]: Why do we need the `async` for?
Can't we do the destructuring in the function declaration instead of the line below?

File: packages/shared/src/hooks/analytics/useAnalyticsQueue.ts [hunk 29]
```diff
@@ -16,7 +16,10 @@ export type PushToQueueFunc = (events: AnalyticsEvent[]) => void;
 
 const ANALYTICS_ENDPOINT = `${apiUrl}/e`;
 
-export default function useAnalyticsQueue(): {
+type UseAnalyticsQueueProps = {
+  method: typeof fetch;
```
Comment (idoshamun) [inline packages/shared/src/hooks/analytics/useAnalyticsQueue.ts:20]: `method` is not explicit enough. Maybe `fetchLike` or `customFetch`, or anything else that hint this is a fetch method

File: packages/shared/src/hooks/analytics/useAnalyticsSharedProps.ts [hunk 30]
```diff
@@ -28,6 +28,7 @@ const getOrGenerateDeviceId = async (): Promise<string> => {
 export default function useAnalyticsSharedProps(
   app: string,
   version: string,
+  deviceId?: string,
```
Comment (idoshamun) [inline packages/shared/src/hooks/analytics/useAnalyticsSharedProps.ts:31]: It is better to not mix two ways of using this hook:
* Getting the device id as an argument
* The hook loads the device id from storage by itself
Due to the companion constraints, we have to use the first method so I suggest applying it throughout the application

**Review by idoshamun — CHANGES_REQUESTED** [review]
The tracking system looks very good but I have a few comments.
I think we also have to track bookmark / remove bookmark, open / close companion sidebar.
If there are any user interactions that we are not tracking in the companion and I missed please let me know. I'm not very aware to the different versions of the companion.

File: packages/shared/src/hooks/analytics/useAnalyticsQueue.ts (hunk 29 above)
Comment (rebelchris) [reply packages/shared/src/hooks/analytics/useAnalyticsQueue.ts:20]: How does `fetchMethod` sound like?
I don't want to make it sound like it's only the proxy as it can also be the "regular" fetch

File: packages/shared/src/hooks/analytics/useAnalyticsSharedProps.ts (hunk 30 above)
Comment (rebelchris) [reply packages/shared/src/hooks/analytics/useAnalyticsSharedProps.ts:31]: I'm a bit stuck what's wisdom here.
If we use option1 how will we pass it?

As the current `getOrGenerateDeviceId` is a promise that returns a string.
I don't want to pre-block from the parent components, nor have the option the deviceId is not set yet.

Perhaps passing a promise as the `deviceId` option could be an option?
That way we can either pass the initial one, or a plain string resolve?

Comment (rebelchris) [reply packages/shared/src/hooks/analytics/useAnalyticsSharedProps.ts:31]: I actually created a new look for the device ID, let me know how that looks.

File: packages/shared/src/hooks/analytics/useAnalyticsQueue.ts (hunk 29 above)
Comment (idoshamun) [reply packages/shared/src/hooks/analytics/useAnalyticsQueue.ts]: Makes sense ☺️

**Comment by idoshamun** [comment]
@rebelchris What's left to be done in terms of development? Maybe we can roll an internal version to try out.
Also, have you checked how to tackle the manifest change thing?

File: packages/extension/src/companion/App.tsx [hunk 31]
```diff
@@ -32,10 +33,15 @@ export default function App({
...
     settings?.optOutCompanion,
   );
+
   if (isOptOutCompanion) {
     return <></>;
   }
 
+  const memoizedFlags = useMemo(() => {
+    return flags;
+  }, [flags]);
```
Comment (idoshamun) [inline packages/extension/src/companion/App.tsx:43]: Just my favorite, not required.

```suggestion
  const memoizedFlags = useMemo(() => flags, [flags]);
```

File: packages/extension/src/companion/App.tsx [hunk 32]
```diff
@@ -44,7 +50,7 @@ export default function App({
       </style>
       <RouterContext.Provider value={router}>
         <QueryClientProvider client={queryClient}>
-          <FeaturesContext.Provider value={{ flags }}>
+          <FeaturesContext.Provider value={{ flags: memoizedFlags }}>
```
Comment (idoshamun) [inline packages/extension/src/companion/App.tsx:53]: The idea of moving to memoized flags was to not create an object on every render.
My suggestion removes the need to create the object every time.

```suggestion
          <FeaturesContext.Provider value={memoizedFlags}>
```

File: packages/extension/src/companion/CompanionMenu.tsx [hunk 33]
```diff
@@ -17,11 +17,22 @@ import { PostBootData } from '@dailydotdev/shared/src/lib/boot';
...
   Modal.setAppElement('daily-companion-app');
 }
 
+const postEventName = (
+  update: Pick<PostBootData, 'upvoted' | 'bookmarked'>,
+): string => {
+  if ('upvoted' in update) {
+    return `${!update.upvoted ? 'remove ' : ''}upvote post`;
+  }
+  return `${!update.bookmarked ? 'remove ' : ''}post bookmark`;
```
Comment (idoshamun) [inline packages/extension/src/companion/CompanionMenu.tsx:33]: We already have these event names somewhere. Maybe we should consider consolidating it somehow?
I'm afraid that when we change something they will not be aligned. This can be a future work as well

**Review by idoshamun — CHANGES_REQUESTED** [review]
Wow! We're almost there 🤯

**Comment by idoshamun** [comment]
There's an issue with the line between the tooltip and the sidebar. It seems to be not aligned

image

**Comment by sshanzel** [comment]
> There's an issue with the line between the tooltip and the sidebar. It seems to be not aligned

I noticed this as well. Thought this was just an isolated case on my local env. Let me log this.

**Comment by sshanzel** [comment]
@rebelchris feel free to review this commit https://github.com/dailydotdev/apps/pull/930/commits/2de70b7e7a44ea8d53a0681ac2dee8a753f99df7 - it is just related to fetching user info for profile tooltip. Thank you!

**Comment by rebelchris** [comment]
@idoshamun @sshanzel I want to see if we can do a last sanity check on this before we merge into master.
This will make further "bugfixing" easier for the companion in my opinion.

File: packages/extension/src/companion/useCompanionActions.ts [hunk 34]
```diff
@@ -0,0 +1,205 @@
...
+      }),
+    {
+      onMutate: () => undefined,
+      onError: (_, __, rollback) => {
+        rollback?.();
+      },
+    },
+  );
+
+  return {
```
Comment (idoshamun) [inline packages/extension/src/companion/useCompanionActions.ts:194]: use memo alert!

**Review by idoshamun — APPROVED** [review]
One minor request. Other than that, we're good to go. May the force be with us

**Comment by sshanzel** [comment]
Have verified locally that fixing the PR conflict was a success. The toast notification works as intended in the companion as well.

cc @rebelchris @idoshamun 

Chrome:
image

Firefox:
image

**Comment by idoshamun** [comment]
Thank you for verifying it @sshanzel !
